File: src/main/optimization-state.js
// src/main/optimization-state.js — 优化项「已应用状态」记账（v2.6.0 P0-1）
// 批次：v2.6.0 优化中心安全增强（借鉴 Pavise SuppressionCore 的 fail-closed 记账不变式）：
//   ① 先记账：执行前写入 pending 记录，写不进去就不改（fail-closed）；
//   ② 执行成功后才把记录转正为 applied；
//   ③ 还原成功才销账：还原失败保留记录，等下次重试。
// 为什么需要它：optimizer-backups.json 只记注册表原值，fsutil 命令 / 服务启停等
// cmd、service 类步骤此前没有任何持久化痕迹；崩溃/重启后无从知道「改过什么」，
// 退役项与半途而废的批量执行会变成永远说不清的历史包袱。
// 数据文件：%APPDATA%\Trim\optimization-state.json（安装版；便携版为程序目录\data，同机制）
// 损坏隔离，与 optimizer-backups 同策略。v2.7.0 起同时承载 detected 段：
//   detected[id] = { optimized: bool, at: ISO } —— 启动全量扫描 + 执行/还原后的及时回写，
//   让「哪些条目已做过优化」跨启动持久留痕（常态化记录），而非只存在渲染层内存里。
const fs = require('fs');
const path = require('path');
const SECURITY = require('./security');

let stateFile = null;
let writeLog = () => {};

function initDataDir(dataDir, logger) {
  stateFile = path.join(dataDir, 'optimization-state.json');
  if (typeof logger === 'function') writeLog = logger;
}

function ready() {
  return !!stateFile;
}

function load() {
  if (!stateFile) return { version: 1, items: {}, detected: {} };
  try {
    const m = JSON.parse(fs.readFileSync(stateFile, 'utf8'));
    if (m && typeof m === 'object' && m.items && typeof m.items === 'object') {
      return {
        version: 1,
        items: m.items,
        detected: (m.detected && typeof m.detected === 'object') ? m.detected : {}
      };
    }
    // 结构不符按损坏处理（与 loadOptBackups 的隔离策略一致）
    throw new Error('结构不符');
  } catch (e) {
    if (fs.existsSync(stateFile)) {
      try {
        const bad = stateFile + '.corrupt-' + Date.now();
        fs.renameSync(stateFile, bad);
        writeLog('warn', `优化状态文件损坏已隔离: ${bad}（${e.message}）`);
      } catch (_) { /* 隔离失败则后续 save 直接覆盖 */ }
    }
    return { version: 1, items: {}, detected: {} };
  }
}

function save(state) {
  if (!stateFile) return false;
  try {
    fs.mkdirSync(path.dirname(stateFile), { recursive: true });
    SECURITY.atomicWriteJson(stateFile, state);
    return true;
  } catch (e) {
    writeLog('error', `写入优化状态失败: ${e.message}`);
    return false;
  }
}

// 读取单条记录（无则 null）
function get(id) {
  return load().items[id] || null;
}

// 执行前记账：写入 pending。返回 false = 写入失败（调用方必须中止执行，fail-closed）。
function recordPending(id, { title = '', kinds = [] } = {}) {
  if (!id) return false;
  const state = load();
  state.items[id] = {
    title: String(title || id),
    appliedAt: new Date().toISOString(),
    kinds: Array.isArray(kinds) ? kinds.filter(k => ['reg', 'cmd', 'service'].includes(k)) : [],
    status: 'pending',        // pending = 已记账但未确认完成（含执行中断/崩溃遗留）
    lastVerify: null,         // pass | partial | unknown（执行后回读验证结果）
    verifiedAt: null
  };
  return save(state);
}

// 执行成功转正：status → applied，并记录回读验证结果
function markApplied(id, verify) {
  if (!id) return false;
  const state = load();
  const rec = state.items[id];
  if (!rec) return false;
  rec.status = 'applied';
  rec.lastVerify = ['pass', 'partial', 'unknown'].includes(verify) ? verify : 'unknown';
  rec.verifiedAt = new Date().toISOString();
  return save(state);
}

// 仅更新回读验证结果（不改状态）
function markVerify(id, verify) {
  if (!id) return false;
  const state = load();
  const rec = state.items[id];
  if (!rec) return false;
  rec.lastVerify = ['pass', 'partial', 'unknown'].includes(verify) ? verify : 'unknown';
  rec.verifiedAt = new Date().toISOString();
  return save(state);
}

// 还原成功后销账。只有调用方确认「还原确实成功」才应调用——失败时保留记录等下次重试。
function remove(id) {
  if (!id) return false;
  const state = load();
  if (!state.items[id]) return true; // 本就不存在，视为已销账
  delete state.items[id];
  return save(state);
}

// 全量清单（供启动扫描 / 状态总览 IPC）
function all() {
  return load().items;
}

// ==================== detected：已优化检测结果持久化（v2.7.0） ====================
// 读取全部检测结果：{ [id]: { optimized, at } }
function getDetectedAll() {
  return load().detected;
}

// 全量替换检测结果（启动扫描完成后调用；只合并写入，不覆盖本函数未涉及的条目以外的段）
function replaceDetected(map) {
  if (!stateFile) return false;
  const state = load();
  state.detected = (map && typeof map === 'object') ? map : {};
  return save(state);
}

// 及时回写单条（执行 pass/partial、还原成功后立即调用，不等下次启动扫描）
function setDetectedEntry(id, optimized) {
  if (!id) return false;
  const state = load();
  state.detected[id] = { optimized: optimized === true, at: new Date().toISOString() };
  return save(state);
}

module.exports = { initDataDir, ready, get, recordPending, markApplied, markVerify, remove, all, getDetectedAll, replaceDetected, setDetectedEntry };

File: src-tauri/src/engine/optimization_state.rs
//! 优化项「已应用状态」记账（对照 src/main/optimization-state.js，v2.6.0 P0-1）。
//!
//! fail-closed 三条不变式：
//!   ① 先记账：执行前写 pending，写不进去就不改系统；
//!   ② 执行成功才转正 applied，并记录回读验证结果；
//!   ③ 还原成功才销账，失败保留记录等下次重试。
//!
//! 数据文件：`<数据目录>/optimization-state.json`，损坏隔离（与 settings 同策略）。
//! v2.7.0 起同时承载 detected 段：{ id: { optimized: bool, at: ISO } }。

use serde_json::{json, Map, Value};

use crate::engine::{log, paths};
use crate::security;

fn state_file() -> std::path::PathBuf {
    paths::app_data_dir().join("optimization-state.json")
}

fn empty_state() -> Value {
    json!({ "version": 1, "items": {}, "detected": {} })
}

fn load() -> Value {
    let f = state_file();
    let v = security::read_json_or_quarantine(&f);
    match v {
        Value::Object(m) if m.get("items").map(|x| x.is_object()).unwrap_or(false) => {
            let mut o = m;
            if !o.get("detected").map(|x| x.is_object()).unwrap_or(false) {
                o.insert("detected".into(), json!({}));
            }
            Value::Object(o)
        }
        _ => empty_state(),
    }
}

fn save(state: &Value) -> bool {
    match security::atomic_write_json(&state_file(), state) {
        Ok(()) => true,
        Err(e) => {
            log::write_log("error", &format!("写入优化状态失败: {e}"));
            false
        }
    }
}

fn iso_now() -> String {
    crate::engine::delete_manifest::iso_now()
}

/// 执行前记账（pending）。false = 写入失败，调用方必须中止（fail-closed）。
pub fn record_pending(id: &str, title: &str, kinds: &[String]) -> bool {
    if id.is_empty() {
        return false;
    }
    let mut state = load();
    let allowed: Vec<Value> = kinds
        .iter()
        .filter(|k| matches!(k.as_str(), "reg" | "cmd" | "service"))
        .map(|k| json!(k))
        .collect();
    if let Some(items) = state.get_mut("items").and_then(|v| v.as_object_mut()) {
        items.insert(
            id.into(),
            json!({
                "title": if title.is_empty() { id } else { title },
                "appliedAt": iso_now(),
                "kinds": allowed,
                "status": "pending",
                "lastVerify": Value::Null,
                "verifiedAt": Value::Null
            }),
        );
    }
    save(&state)
}

/// 执行后转正（applied）+ 回读验证三态
pub fn mark_applied(id: &str, verify: &str) -> bool {
    let v = normalize_verify(verify);
    let mut state = load();
    let Some(rec) = state
        .get_mut("items")
        .and_then(|i| i.as_object_mut())
        .and_then(|m| m.get_mut(id))
        .and_then(|r| r.as_object_mut())
    else {
        return false;
    };
    rec.insert("status".into(), json!("applied"));
    rec.insert("lastVerify".into(), json!(v));
    rec.insert("verifiedAt".into(), json!(iso_now()));
    save(&state)
}

/// 还原成功销账（本就不存在视为成功）
pub fn remove(id: &str) -> bool {
    if id.is_empty() {
        return false;
    }
    let mut state = load();
    let Some(items) = state.get_mut("items").and_then(|v| v.as_object_mut()) else {
        return true;
    };
    if !items.contains_key(id) {
        return true;
    }
    items.remove(id);
    save(&state)
}

/// 全部记账条目
pub fn all() -> Map<String, Value> {
    load()
        .get("items")
        .and_then(|v| v.as_object())
        .cloned()
        .unwrap_or_default()
}

/// 单条记账（无则 null）
pub fn get(id: &str) -> Option<Value> {
    load().get("items")?.get(id).cloned()
}

/// 回写单条检测结果
pub fn set_detected_entry(id: &str, optimized: bool) -> bool {
    if id.is_empty() {
        return false;
    }
    let mut state = load();
    if let Some(d) = state.get_mut("detected").and_then(|v| v.as_object_mut()) {
        d.insert(
            id.into(),
            json!({ "optimized": optimized, "at": iso_now() }),
        );
    }
    save(&state)
}

/// 读取全部检测结果
pub fn detected_all() -> Map<String, Value> {
    load()
        .get("detected")
        .and_then(|v| v.as_object())
        .cloned()
        .unwrap_or_default()
}

fn normalize_verify(v: &str) -> &'static str {
    match v {
        "pass" => "pass",
        "partial" => "partial",
        _ => "unknown",
    }
}
